Hydra optuna sweeper custom search space 1909

MattiasDC opened the pull request:
<!-- Thank you for sending a PR and taking the time to improve Hydra -->

## Motivation
This pull request addresses [issue 1906](https://github.com/facebookresearch/hydra/issues/1906). It introduces a new configuration option for the Hydra-Optuna-Sweeper plugin that allows extending Optuna trial objects via Pythonic search spaces.

For a project I would like to use Hydra and Optuna as a replacement of current customly written code. But in order to do that I need to have the possibility to do some complex trial configuration that is currently not supported by the Optuna plugin. With this extension we can easily combine the strength of Hydra with Optuna.

### Have you read the [Contributing Guidelines on pull requests]
Yes

## Test Plan
I've added unit tests and added an extra example that shows how to use this added functionality.

## Related Issues and PRs
 [issue 1906](https://github.com/facebookresearch/hydra/issues/1906)

MattiasDC commented on plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:
```diff
@@ -147,38 +184,56 @@ def setup(
+            for param_name, distribution in search_space.items():
+                trial._suggest(param_name, distribution)
+            for param_name, value in fixed_params.items():
+                trial.set_user_attr(param_name, value)
```
Setting fixed parameters as user_attrs in Optuna was previously not done. I think this however makes it more clear for Optuna that this parameter exists and future runs with it not being fixed, might use reuse the information of runs.

MattiasDC commented on plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:
```diff
@@ -105,6 +109,35 @@ def create_optuna_distribution_from_override(override: Override) -> Any:
     raise NotImplementedError(f"{override} is not supported by Optuna sweeper.")
 
+def get_function_from_path(import_string: str) -> Callable[[Trial], None]:
+    method_path = import_string.split(".")
```
Not an expert on imports, but I hope this doesn't bring any security issues that we need to take into account (e.g. ../../../etc/passwd). Since we split on '.', I think this is safe?

MattiasDC commented on plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:
```diff
@@ -109,22 +119,34 @@ def create_optuna_distribution_from_override(override: Override) -> Any:
 
-def get_function_from_path(import_string: str) -> Callable[[Trial], None]:
+
+def get_function_from_path(import_string: str) -> Any:
```
I had to use 'Any' in the end since 'Callable' could not be used due to linting errors of it not being fully specified. Writing Callable[[DictConfig, Trial], None] doesn't work as mypy cannot defer the type from the getattr call. Furthermore 'isinstance(function, Callable[[DictConfig, Trial], None])' is not yet supported in Python. Generics can't be used in 'isinstance'

Jasha10 replied:
> 'isinstance(function, Callable[[DictConfig, Trial], None])' is not yet supported...

FYI there is a builtin [`callable`](https://docs.python.org/3/library/functions.html#callable) function that can do part of the job :)

Jasha10 replied:
> 'isinstance(function, Callable[[DictConfig, Trial], None])' is not yet supported...

FYI there is a builtin callable function that can do part of the job :)

MattiasDC replied:
Thanks for the info! Indeed, I could reused the get_method :)

Jasha10 commented:
Hi @MattiasDC, thank you very much for your contribution! We will take a look at it shortly :)

Jasha10 commented on plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:
```diff
@@ -147,38 +206,56 @@ def setup(
         self.sweep_dir = config.hydra.sweep.dir
 
+    def _get_directions(self) -> List[str]:
+        directions: List[str]
```
I think this line can be deleted.
```suggestion
```

Jasha10 reviewed (commented):
A few initial comments:

Jasha10 commented:
Thanks for updating the PR, @MattiasDC.
@jieru-hu and I will need a few days to review this.

Jasha10 commented on website/docs/plugins/optuna_sweeper.md:
```diff
@@ -299,3 +299,52 @@ python example/multi-objective.py --multirun
 
 ![Pareto-optimal solutions](/plugins/optuna_sweeper/multi_objective_result.png)
+
+## Example 3:  Custom-Search-Space Optimization
```
```suggestion
## EXPERIMENTAL:  Custom-Search-Space Optimization
```

Jasha10 commented on website/docs/plugins/optuna_sweeper.md:
````diff
@@ -299,3 +299,52 @@ python example/multi-objective.py --multirun
+max_z_difference_from_x: 0.5
+```
+
+The `custom_search_space` configuration option refers to the `configure` function in `example/custom-search-space-objective.py`. This method accepts both a DictConfig with already set options and the trial object which needs further configuration. In this example we limit `z` the difference between `x` and `z` to be no more than 0.5.
````
```suggestion
The `custom_search_space` configuration option refers to the `configure` function in `example/custom-search-space-objective.py`. This method accepts both a DictConfig with already set options and the trial object which needs further configuration. In this example we limit `z` the difference between `x` and `z` to be no more than 0.5.
Note that this `custom_search_space` API should be considered experimental and is subject to change.
```

Jasha10 reviewed (changes requested):
We might want to change this API in the future; let's call it "experimental" for now.

Jasha10 commented on plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:
```diff
@@ -209,7 +209,7 @@ def _configure_trials(
             params = dict(trial.params)
             params.update(fixed_params)
-            overrides.append(tuple(f"{name}={val}" for name, val in params.items()))
+            overrides.append(tuple(f"++{name}={val}" for name, val in params.items()))
```
This change has caused CI to fail. I can reproduce the issue locally.
The problem is that we might have e.g. `name == "~z", as was the case with the failing test. In that case the f-string `f"++{name}"` would be equal to `"++~z"`, which is no bueno.

I am not convinced that adding `++` is the right move here.

MattiasDC replied:
I changed it back to remove '++', however I also extended the test to cover a case that I feel that must be supported (you can't define all parameters in advance in the DictConfig). Note that in the committed change 'w' cannot be overridden, because it's not known.

MattiasDC replied:
Any ideas on how to resolve the currently failing test?

Jasha10 replied:
I will take a look...

MattiasDC replied:
@Jasha10 I've adapted the code to only add a '+' if the custom search space introduced a new parameter. Instead of by default add '++' to all parameters

Jasha10 commented on plugins/hydra_optuna_sweeper/news/1906.feature:
```diff
@@ -0,0 +1 @@
+Add 'custom_search_space' configuration option to allow extending trial objects programmatically
```
```suggestion
Add experimental 'custom_search_space' configuration node to allow extending trial objects programmatically.
```

I think the word "option" could be confusing because it has a different meaning in the context of Hydra's defaults list (see e.g. ["Terminology - Config Group Option"](https://hydra.cc/docs/advanced/terminology/#config-group-option)).

MattiasDC commented:
Hi @Jasha10,
Not trying to push you here, but I'm wondering whether you can give me an ETA? :)

Jasha10 commented:
Hi @MattiasDC, thanks for the nudge :)

Jasha10 commented on plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:
```diff
@@ -147,38 +179,71 @@ def setup(
+
+                # If the added a trial parameter using custom search space that is not part of the config yet,
+                # we have to add it as a new config value
+                return f"+{override}"
```
I am still not comfortable with this; it breaks consistency with Hydra's command-line API as well as with the precedent from PR #1886.
Hydra should fail fast if someone tries to set a parameter that does not exist in the config ... *unless* if the user explicitly adds a plus `+` to signal that adding the new parameter is intended.

This being said, I agree with you that it should be possible to use `custom_search_space` to add a new parameter to the config. Here is my proposal:
The user's custom search space function should call `trial.suggest_float` using a plus symbol for the name of the suggested parameter. For example:
```python
# custom-search-space-objective.py

...

def configure(cfg: DictConfig, trial: Trial) -> None:
    x_value = trial.params["x"]
    trial.suggest_float(
        "z",
        x_value - cfg.max_z_difference_from_x,
        x_value + cfg.max_z_difference_from_x,
    )
    trial.suggest_float("+w", 0.0, 1.0)  # note +w here, not w

...
```

MattiasDC replied:
Changed it accordingly

Jasha10 commented:
Overall this PR looks good. Could you please rebase against Hydra's `main` branch to resolve the merge conflict?

MattiasDC commented:
Hi @Jasha10 ,
I've addressed your remark and I've merged from main and resolved the conflicts with the new style params.
I also simplified it a bit further.
While doing the rebase I've messed up, so the commits might seem 'new' again as I had to do a force push.
My apologies for this.

I have 4 'failing' tests, but these seem to be flaky, is this a known problem?
They are related to jupyter_notebook and brew environment

Jasha10 commented:
Thanks Mattias!

> I have 6 'failing' tests, but these seem to be flaky, is this a known problem? They are related to jupyter_notebook and brew environment

Yes, the jupyter notebook failures are a known issue that motivated PR #2123. Incorporating that diff into this branch will (hopefully) solve the issue.

Jasha10 commented on plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:
```diff
@@ -115,6 +126,23 @@ def create_optuna_distribution_from_override(override: Override) -> Any:
     raise NotImplementedError(f"{override} is not supported by Optuna sweeper.")
 
 
+def create_params_from_overrides(
+    arguments: List[str],
+) -> Tuple[Dict[str, BaseDistribution], Dict[str, Any]]:
+    parser = OverridesParser.create()
+    parsed = parser.parse_overrides(arguments)
+    search_space = dict()
```
In `OptunaSweeperImpl.__init__`, the `search_space` variable has type `Optional[DictConfig]`, but in this function `search_space` has type `Dict[str, BaseDistribution]`. This type mismatch is confusing... Could you please rename `search_space` here to a different name?

MattiasDC replied:
Changed it to search_space_distributions

Jasha10 commented on plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:
```diff
@@ -187,6 +217,41 @@ def setup(
+    def _configure_trials(
+        self,
+        trials: List[Trial],
+        search_space: Dict[str, BaseDistribution],
```
Same remark as before about renaming `search_space`.

MattiasDC replied:
Changed it to search_space_distributions

Jasha10 commented on plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:
```diff
@@ -187,6 +217,41 @@ def setup(
+                self.custom_search_space(self.config, trial)
+
+            overlap = trial.params.keys() & trial.user_attrs
+            assert len(overlap) == 0, (
+                "Overlapping fixed parameters and search space parameters found!"
+                f"Overlapping parameters: {list(overlap)}"
+            )
```
Can this error be triggered by the user making a mistake with their config? If so, this should be a `ValueError` instead of an `AssertionError`.

MattiasDC replied:
Yes it can, I will change this

Jasha10 commented on plugins/hydra_optuna_sweeper/example/custom-search-space/config.yaml:
```diff
@@ -0,0 +1,23 @@
+    n_trials: 20
+    n_jobs: 1
+
+    params:
+      x: range(-5.5, 5.5, 0.5)
+      y: choice(-5, 0, 5)
+    custom_search_space: .custom-search-space-objective.configure
+
```
```suggestion
    params:
      x: range(-5.5, 5.5, 0.5)
      y: choice(-5, 0, 5)
    # `custom_search_space` should be a dotpath pointing to a
    # callable that provides search-space configuration logic:
    custom_search_space: .custom-search-space-objective.configure

```

Jasha10 commented on website/docs/plugins/optuna_sweeper.md:
````diff
@@ -252,3 +252,53 @@ python example/multi-objective.py --multirun
+```yaml
+defaults:
+  - override hydra/sweeper: optuna
+  - override hydra/sweeper/sampler: tpe
````
```suggestion
```

Jasha10 commented on website/docs/plugins/optuna_sweeper.md:
```diff
@@ -252,3 +252,53 @@ python example/multi-objective.py --multirun
+    n_trials: 20
+    n_jobs: 1
+
+    search_space:
+      x:
+        type: float
+        low: -5.5
+        high: 5.5
+        step: 0.5
+      y:
+        type: categorical
+        choices: [-5, 0, 5]
+    custom_search_space: example.custom-search-space-objective.configure
```
```suggestion
    params:
      x: range(-5.5, 5.5, 0.5)
      y: choice(-5, 0, 5)
    # `custom_search_space` should be a dotpath pointing to a
    # callable that provides search-space configuration logic:
    custom_search_space: .custom-search-space-objective.configure
```

Jasha10 commented on website/docs/plugins/optuna_sweeper.md:
````diff
@@ -252,3 +252,53 @@ python example/multi-objective.py --multirun
+z: 100
+max_z_difference_from_x: 0.5
+```
+
````
````suggestion
```python
# example/custom-search-space-objective.py

...

def configure(cfg: DictConfig, trial: Trial) -> None:
    x_value = trial.params["x"]
    trial.suggest_float(
        "z",
        x_value - cfg.max_z_difference_from_x,
        x_value + cfg.max_z_difference_from_x,
    )
    trial.suggest_float("+w", 0.0, 1.0)  # note +w here, not w as w is a new parameter

...
```

````

Jasha10 commented on website/docs/plugins/optuna_sweeper.md:
```diff
@@ -252,3 +252,53 @@ python example/multi-objective.py --multirun
+
+## EXPERIMENTAL:  Custom-Search-Space Optimization
+
+In the same directory, `example/custom-search-space-objective.py` implements a simple benchmark function, which needs to be minimized. The example shows the use of Optuna's [pythonic search spaces](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/002_configurations.html) in combination with Hydra. Part of the search space configuration is defined in config files, part of it is wri ...
```
```suggestion
You can find a full example in the same directory as before, where `example/custom-search-space-objective.py` implements a benchmark function to be minimized.
The example shows the use of Optuna's [pythonic search spaces](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/002_configurations.html) in combination with Hydra.
Part of the search space configuration is defined in config files, and part of it is written in Python.
```

Jasha10 commented on website/docs/plugins/optuna_sweeper.md:
```diff
@@ -252,3 +252,53 @@ python example/multi-objective.py --multirun
+
+In the same directory, `example/custom-search-space-objective.py` implements a simple benchmark function, which needs to be minimized. The example shows the use of Optuna's [pythonic search spaces](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/002_configurations.html) in combination with Hydra. Part of the search space configuration is defined in config files, part of it is wri ...
+
+<details><summary>Configuration of the custom-search-space-objective optimization example</summary>
```
```suggestion
<details><summary>Example: Custom search space configuration</summary>
```

Jasha10 commented on website/docs/plugins/optuna_sweeper.md:
```diff
@@ -252,3 +252,53 @@ python example/multi-objective.py --multirun
 
 ![Pareto-optimal solutions](/plugins/optuna_sweeper/multi_objective_result.png)
+
+## EXPERIMENTAL:  Custom-Search-Space Optimization
+
```
```suggestion
## EXPERIMENTAL:  Custom-Search-Space Optimization

Hydra's Optuna Sweeper allows users to provide a hook for custom search space configuration.
This means you can work directly with the `optuna.trial.Trial` object to suggest parameters.
To use this feature, define a python function with signature `Callable[[DictConfig, optuna.trial.Trial], None]`
and set the `hydra.sweeper.custom_search_space` key in your config to target that function.

```

Jasha10 commented on website/docs/plugins/optuna_sweeper.md:
````diff
@@ -252,3 +252,53 @@ python example/multi-objective.py --multirun
+max_z_difference_from_x: 0.5
+```
+
+The `custom_search_space` configuration option refers to the `configure` function in `example/custom-search-space-objective.py`. This method accepts both a DictConfig with already set options and the trial object which needs further configuration. In this example we limit `z` the difference between `x` and `z` to be no more than 0.5.
+Note that this `custom_search_space` API should be considered experimental and is subject to change.
+</details>
````
```suggestion
</details>

The method that `custom_search_space` points to must accepts both a DictConfig with already set options and a trial object which needs further configuration. In this example we limit `z` the difference between `x` and `z` to be no more than 0.5.
Note that this `custom_search_space` API should be considered experimental and is subject to change.
```

Jasha10 commented on website/docs/plugins/optuna_sweeper.md:
```diff
@@ -252,3 +252,53 @@ python example/multi-objective.py --multirun
+
+### Order of trial configuration
+Configuring a trial object is done in the following sequence:
+  - `search_space` parameters from the config are set
```
```suggestion
  - search space parameters are set from the `hydra.sweeper.params` key in the config
```

Jasha10 reviewed (changes requested):
The implementation looks good.
See my comments below:

Jasha10 commented on plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py:
```diff
@@ -208,6 +235,32 @@ def _dominates(values_x: List[float], values_y: List[float]) -> bool:
+    assert isinstance(returns, DictConfig)
+    assert returns.name == "optuna"
+    assert (
+        returns["best_params"]["x"] - returns["best_params"]["z"]
```
```suggestion
        abs(returns["best_params"]["x"] - returns["best_params"]["z"])
```

MattiasDC commented on plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py:
```diff
@@ -208,6 +235,32 @@ def _dominates(values_x: List[float], values_y: List[float]) -> bool:
+    w = returns["best_params"]["+w"]
+    assert w >= 0
+    assert w <= 1
+
```
Thanks, learned something new today! :)

MattiasDC commented:
> Thanks Mattias!
> 
> > I have 6 'failing' tests, but these seem to be flaky, is this a known problem? They are related to jupyter_notebook and brew environment
> 
> Yes, the jupyter notebook failures are a known issue that motivated the recent PR #2123. Incorporating that diff into this branch will (hopefully) solve the issue.

Jupyter issues seem to be fixed. Brew issues are not. Is there any way to rerun the checks?

Jasha10 commented:
I've just triggered a rerun.
> Is there any way to rerun the checks?

You can always push an empty commit to the PR branch :P

MattiasDC commented:
> I've just triggered a rerun.
> 
> > Is there any way to rerun the checks?
> 
> You can always push an empty commit to the PR branch :P

Seemed to do the trick!

Jasha10 commented on plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:
```diff
@@ -115,6 +127,23 @@ def create_optuna_distribution_from_override(override: Override) -> Any:
+) -> Tuple[Dict[str, BaseDistribution], Dict[str, Any]]:
+    parser = OverridesParser.create()
+    parsed = parser.parse_overrides(arguments)
+    search_space = dict()
```
```suggestion
    search_space_distributions = dict()
```

Jasha10 commented on plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:
```diff
@@ -115,6 +127,23 @@ def create_optuna_distribution_from_override(override: Override) -> Any:
+        param_name = override.get_key_element()
+        value = create_optuna_distribution_from_override(override)
+        if isinstance(value, BaseDistribution):
+            search_space[param_name] = value
```
```suggestion
            search_space_distributions[param_name] = value
```

Jasha10 commented on plugins/hydra_optuna_sweeper/hydra_plugins/hydra_optuna_sweeper/_impl.py:
```diff
@@ -115,6 +127,23 @@ def create_optuna_distribution_from_override(override: Override) -> Any:
+            search_space[param_name] = value
+        else:
+            fixed_params[param_name] = value
+    return search_space, fixed_params
```
```suggestion
    return search_space_distributions, fixed_params
```

Jasha10 reviewed (approved):
Looks good to me!

MattiasDC commented:
Hi @Jasha10,

Thanks for your time and patience.
This was my first open source pull request, so if you have any feedback, feel free to leave some! :)

MattiasDC commented:
Hydra Optuna sweeper plugin now allows to extend the trial object programmatically using a custom search space configuration option.

MattiasDC commented:
Also, it seems I don't have write access to merge this PR, so I leave that up to you @Jasha10

Jasha10 commented:
> Hi @Jasha10,
> 
> Thanks for your time and patience. This was my first open source pull request, so if you have any feedback, feel free to leave some! :)

Wow, well you did a nice job with it! It was good collaborating with you 😄

@jieru-hu, could you please give this a final review before we merge?

jieru-hu commented:
looks really good! thanks @MattiasDC !
